diarymanagement-code-r273/tbaclient/tags/final.rc1/tbaclient/src/TBA/Communications/ServerCom.java

```diff
@@ -0,0 +1,834 @@
+package TBA.Communications;
+
+import TBA.Exceptions.*;
+import TBA.Messages.*;
+import TBA.Data.*;
+import TBA.Security.PasswordHashing;
+import java.io.*;
+import java.util.Calendar;
+import java.util.Vector;
+import javax.net.ssl.SSLSocket;
+import javax.net.ssl.SSLSocketFactory;
+import java.util.logging.Logger;
+
+/**
+ * This class handles the connection to the TBA server
+ *<p>
+ *
+ * <p>
+ * @author Dan McGrath
+ *
+ * @version $Rev:: 244           $ $Date:: 2009-10-22 #$
+ */
+
+public class ServerCom
+{
+   private SSLSocket myConx = null;
+   private int socketPort = 0;
+   private String socketAddress = null;
+   private DataOutputStream os;
+   private DataInputStream is;
+   private boolean isConnected = false;
+   private DBServer DBCom;
+   private String lastError = null;
+
+   /** LOGIT field provides 4 levels of feedback to user info, fine, finer , severe*/
+   public final static Logger LOGIT = Logger.getLogger(ServerCom.class.getName());
+
+   /**
+    * This returns the error message response from the last server command,
+    * if there was one, or null if there was not. It is cleared every time a
+    * server command is called, including when calling this method.
+    * @return ret - error message or null if none
+    */
+   public String getLastError()
+   {
+      String ret = null;
+      if(lastError != null)
+      {
+         ret = new String(lastError);
+      }
+      lastError = null;
+
+      return ret;
+   }
+
+   /**
+    * This static member is used to extract the SSL key store from within the
+    * JAR file.
+    */
+   public static void extractKeyStore()
+   {
+      InputStream keyfile = ServerCom.class.getResourceAsStream("tbaKeyStore");
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer

**Problem**
This line of code contains a resource that might not be closed properly. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the following resource: *keyfile*. The resource is referenced in statements at the following line: 72. The resource closure statement is at line: 76. There are other execution paths that do not contain closure statements, e.g., when exception is thrown by InputStream.read. Either a) close *keyfile* in a try-finally block or b) close the resource by declaring *keyfile* in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).